README.md
# substrate-starter v0.91

arbitrary number of interconnected substrate nodes, incl key generation, chainspec, etc.

## entry points
main scripts, what they do, and how often you need them:

* [start.sh](start.sh): substrate-starter (often)
* [config.sh](config.sh): paths, definitions, etc (seldom/never)
* [init.sh](init.sh) : install, compile, etc (once)

### dependencies

This new 2020 script collection does NOT work with older node-template / substrate anymore; they swapped `babe` for `aura`, and that results in slight changes to the chainspec format. So use this node version: [node-template 8b6fe6666d](https://github.com/substrate-developer-hub/substrate-node-template/tree/8b6fe6666d589486dd541663a32ffd98f2e21d74) (which is on the basis of [substrate 3e651110a](https://github.com/paritytech/substrate/tree/3e651110aa06aa835790df63410a29676243fc54)).

Work through the [installation manual](https://substrate.dev/docs/en/overview/getting-started) so that these are available on the `$PATH`:

    node-template --version; subkey --version
    
> node-template 2.0.0-8b6fe66-x86_64-linux-gnu  
> subkey 2.0.0  

The Python library `websockets` is needed, and now gets installed in [./start.sh 3](start.sh) in the first called script: [source scripts/dependencies.sh](source scripts/dependencies.sh). If you want to call any of the modular [scripts](scripts) alone, make sure to have that same virtualenv enabled: `source env/bin/activate`.


### how to start

    ./start.sh 3
    
Then read everything. If errors, then fix from the top. The slightest error anywhere will cause it to fail later, for sure.

    ps aux | grep node-template
    pkill node-template
    
more infos given at the end of the script. Especially note the

> And watch the logs with: ...

## Memory
I have tested with increasing number of nodes:

| number of interconnected nodes | Total memory used | memory per node in MB |
|--------------------------------|-------------------|-----------------------|
| 10                             | 833               | 83                    |
| 20                             | 1692              | 85                    |
| 30                             | 2579              | 86                    |
| 40                             | 3595              | 90                    |
| 50                             | 4697              | 94                    |
| 80                             | 8518              | 106                   |
| 99                             | 11015             | 111                   |

Stability: 30 nodes, on a 2018 Laptop, run just fine. No error message whatsoever. And mostly `(29 peers)`!  

However when starting larger numbers of nodes (on a too small machine), checking the log files, you will see error messages like these:

* Handler initialization process is too long with PeerId("QmRvr...")
* Polling the network future took 35s
* Essential task failed. Shutting down service.

Especially the last one probably means that also the memory measurements become incorrect then? All this could be fixed somehow probably, but I don't care enough.

To start that many nodes on one machine is a rather academic extreme case anyways. It does NOT make sense to start more than a couple of nodes. But for that - these scripts are very useful, right?


## TODO:
Installation script `./init.sh` to prepare a fresh machine. For now, just install `node-template` and `subkey` manually, please.

When starting 55 nodes or more (lol), the port 9955 is taken already.  

Dockerize:

* Debian based
* run init.sh
* entrypoint start.sh
* map ports 30333, 9944, 9933 of the bootnode


## Alice Bob Charlie Dave
The built-in keys allow a simple local network too:

```
scripts/ABCD_start-4-nodes.sh
tail -f logs/ABCD_alice.log logs/ABCD_bob.log logs/ABCD_charlie.log logs/ABCD_dave.log
pkill node-template
scripts/ABCD_purge-4-chains.sh
```
Esp. if you are new to this, that approach is easiest to understand. But no control over keys, chainspec, etc. And it looks as if always only Alice and Bob are taking turns in authoring blocks ("Starting consensus session on top of parent ..."), while Charlie and Dave always only sync ("Imported ..."):

```
tail -f logs/ABCD_charlie.log logs/ABCD_dave.log
tail -f logs/ABCD_alice.log logs/ABCD_bob.log
```

## Credits
![logo grants badge](img/web3_foundation_grants_badge_black_smaller.png)

This little side project for adapting [chainhammer](https://github.com/drandreaskrueger/chainhammer) to substrate was financed by the web3foundation.

CHANGELOG
CHANGELOG

v0.91
* websockets installed in virtualenv, so no version conflicts with other projects on the same machine
  * new scripts/dependencies.sh does that.
  * explanations in README.md: source env/bin/activate


v0.9
extended from 9 to 99 nodes maximum:
* changed ports intervall from 9933... to 9800-9899 and 9944... to 9900-9999
* filenames seed7.secret --> seed07.secret
* start all nodes with 'nice' to not freeze the machine when many nodes
* README.md
  * memory measurements for 10,20,30,40,50,80,99 nodes 
  * mentioned the error messages when huge number of nodes
  * manual for node-template installation linked
* added the web3foundation grants badge


v0.8
works with node-template-2.0.0-8b6fe66 ( = substrate 2.0.0-3e651110a ) 
* aura+grandpa consensus
* replaced babe with aura everywhere
* See https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/issues/477
* fixed different list types in subkey_addresses.sh: aura addresses have not weights, grandpa addresses have weight 1
* remove bootNodes from .json and -raw.json
* Python first checks whether websockets is installed, or fails with help message
* much more infos in README.md


v0.7
* worked well with node-template-2.0.0-7d7e74fb7 
* babe+grandpa consensus


< v0.7
* see commit history


config.sh
#!/bin/bash

# folder paths etc, to be loaded from every script.

VERSION=v0.91

# binary to use, e.g. node-template, or substrate
SUBSTRATE=node-template

# keys and chainspec are stored in folder 
GENERATED=generated

# logfiles go into this folder
LOGFILES=logs

# all scripts live in folder
SCRIPTS=scripts

# data folders for the nodes
BASEPATH=/tmp

# naming of the nodes begins with
FILESTUB=SS

# build-spec --chain local  
TEMPLATE="local"

# my own chainspec is named:
CHAIN="$GENERATED/$FILESTUB"-raw.json

# CLI args for all nodes: 
CONFIG="--validator --chain "$CHAIN" --telemetry-url ws://telemetry.polkadot.io:1024"

# ports of the FIRST node, subsequent nodes are incremented+1
PORT_P2P=30333
# original port intervall allowed only 10 nodes: 
#PORT_RPC=9933 # default
#PORT_WS=9944 # default
# made port intervall 100 nodes wide:
PORT_RPC=9800 
PORT_WS=9900

# first arg must be number of nodes 
checknum () {
    
    NUM=$1
    if [ -z "$NUM" ] 
    then
        echo "give number of nodes as integer arg please"
        exit
    fi
    
    if [ "$NUM" -ge 100 ] 
    then
        echo "max number of nodes is 99, then it would need some tiny code upgrades:"
        echo e.g.
        echo rpc $PORT_RPC + 100 = ws $PORT_WS port collision
        echo log file names 01 .. 99 good ... 100 problem
        exit
    fi
}

settrap () {
    # exit when any command fails
    set -e
    # keep track of the last executed command
    trap 'last_command=$current_command; current_command=$BASH_COMMAND' DEBUG
    # echo an error message before exiting
    trap 'echo; echo "\"${last_command}\" command filed with exit code $?."' EXIT
}

chapter () {
    echo 
    echo ==========================================
    echo = $1
    echo ==========================================
        
    $1
}

NumberWithLeadingZero () {
	if [ "$1" -ge 100 ] 
	then
		echo "Not implemented for > 99. EXITING."
		exit 128
	fi
	Number=$1
	if [ "$Number" -lt 10 ] 
	then
		Number="0"$Number
	fi
}

testNumberWithLeadingZero () {
	# bash -c "source config.sh; testNumberWithLeadingZero"
	NumberWithLeadingZero 1
	echo $Number
	NumberWithLeadingZero 10
	echo $Number
	NumberWithLeadingZero 100
}
